cub/cub/device/dispatch/dispatch_streaming_reduce.cuh

```diff
@@ -0,0 +1,444 @@
+{
+public:
+  using super_t = thrust::iterator_adaptor<offset_iterator<Iterator, OffsetItT>, Iterator>;
+
+  _CCCL_HOST_DEVICE _CCCL_FORCEINLINE offset_iterator(const Iterator& it, OffsetItT offset_it)
+      : super_t(it)
+      , offset_it(offset_it)
+  {}
+
+  // befriend thrust::iterator_core_access to allow it access to the private interface below
+  friend class thrust::iterator_core_access;
+
+private:
+  OffsetItT offset_it;
+
+  _CCCL_DEVICE _CCCL_FORCEINLINE typename super_t::reference dereference() const
+  {
+    return *(this->base() + (*offset_it));
+  }
+};
+
+template <typename Iterator, typename OffsetItT>
+_CCCL_HOST_DEVICE offset_iterator<Iterator, OffsetItT> make_offset_iterator(Iterator it, OffsetItT offset_it)
+{
+  return offset_iterator<Iterator, OffsetItT>{it, offset_it};
+}
+
+/**
+ * Helper transform operation that writes to the (index, extremum)-key-value-pair to the user-provided output iterator.
+ * This helper allows "implicit conversion" between KeyValuePair types that have a different key type, which may happen
+ * if the user uses a different index type than the global offset type used by the algorithm.
+ */
+template <typename KeyValuePairOutItT>
+struct write_to_user_out_it
+{
+  KeyValuePairOutItT out_it;
+
+  template <typename IndexT, typename OffsetT, typename ResultT>
+  _CCCL_DEVICE _CCCL_FORCEINLINE void operator()(IndexT, cub::KeyValuePair<OffsetT, ResultT> reduced_result)
+  {
+    using kv_pair_t = ::cuda::std::_If<
+      ::cuda::std::is_assignable_v<decltype(*std::declval<KeyValuePairOutItT>()),
+                                   cub::KeyValuePair<::cuda::std::uint64_t, ResultT>>,
+      cub::KeyValuePair<::cuda::std::uint64_t, ResultT>,
+      ::cuda::std::_If<::cuda::std::is_assignable_v<decltype(*std::declval<KeyValuePairOutItT>()),
+                                                    cub::KeyValuePair<::cuda::std::int64_t, ResultT>>,
+                       cub::KeyValuePair<::cuda::std::int64_t, ResultT>,
+                       ::cuda::std::_If<::cuda::std::is_assignable_v<decltype(*std::declval<KeyValuePairOutItT>()),
+                                                                     cub::KeyValuePair<::cuda::std::uint32_t, ResultT>>,
+                                        cub::KeyValuePair<::cuda::std::uint32_t, ResultT>,
+                                        cub::KeyValuePair<::cuda::std::int32_t, ResultT>>>>;
+
+    using index_t = typename kv_pair_t::Key;
+    *out_it       = kv_pair_t{static_cast<index_t>(reduced_result.key), reduced_result.value};
+  }
+};
+
+/******************************************************************************
+ * Single-problem streaming reduction dispatch
+ *****************************************************************************/
+
+/**
+ * @brief Utility class for dispatching the appropriately-tuned kernels for
+ *        device-wide reduction
```

Comment:
IIRC, @miscco told me documentation comments must use `//!`. `/**` works for doxygen, but IIRC there is some problem with Sphinx.